activesupport/lib/active_support/duration/iso8601_duration_parser.rb

```diff
@@ -0,0 +1,61 @@
+# Parses a string formatted according to ISO 8601 Duration into the hash .
+#
+# See http://en.wikipedia.org/wiki/ISO_8601#Durations
+# Parts of code are taken from ISO8601 gem by Arnau Siches (@arnau).
+# This parser isn't so strict and allows negative parts to be present in pattern.
+class ActiveSupport::Duration::ISO8601DurationParser
+  attr_reader :parts
+
+  class ParsingError < ::StandardError; end
+
+  def initialize(iso8601duration)
+    @raw = iso8601duration
+    parse_iso_duration!
+    construct_parts
+    validate_parts!
+  end
+
+  private
+
+  def parse_iso_duration!
+    @match = @raw.match(/^
+                (?<sign>\+|-)?
+                P(?:
+                  (?:
+                    (?:(?<years>-?\d+(?:[,.]\d+)?)Y)?
+                    (?:(?<months>-?\d+(?:[.,]\d+)?)M)?
+                    (?:(?<days>-?\d+(?:[.,]\d+)?)D)?
+                    (?<time>T
+                      (?:(?<hours>-?\d+(?:[.,]\d+)?)H)?
+                      (?:(?<minutes>-?\d+(?:[.,]\d+)?)M)?
+                      (?:(?<seconds>-?\d+(?:[.,]\d+)?)S)?
+                    )?
+                  ) |
+                  (?<weeks>-?\d+(?:[.,]\d+)?W)
+                ) # Duration
+              $/x) || raise(ParsingError.new("Invalid ISO 8601 duration: #{@raw}"))
+  end
+
+  # Constructs parts compatible with +ActiveSupport::Duration+ ones.
+  def construct_parts
+    sign = @match[:sign] == '-' ? -1 : 1
+    @parts = @match.names.zip(@match.captures).reject { |_k,v| v.nil? }.map do |k, v|
+      value = /\d+[\.,]\d+/ =~ v ? v.sub(',', '.').to_f : v.to_i
```

Comment:
This is to satisfy the spec. As referenced by wikipedia in [ISO 8601](http://en.wikipedia.org/wiki/ISO_8601#cite_note-15):

> ISO 8601:2004(E), ISO, 2004-12-01, 4.2.2.4 ... the decimal fraction shall be divided from the integer part by the decimal sign specified in ISO 31-0, i.e. the comma [,] or full stop [.]. Of these, the comma is the preferred sign.

There is no place for `I18n` because format is not locale dependent